### PR TITLE
Add new CCCL library for CUDA

### PR DESCRIPTION
**Pull request by jrhemstad**
The Thrust/CUB/libcudacxx repositories were recently merged to https://github.com/nvidia/cccl.

This PR adds a new `CCCL` library for the CUDA language and marks the old Thrust/CUB/libcudacxx libraries as legacy. 

See also: https://github.com/compiler-explorer/infra/pull/1090

### CODE REVIEW

File: etc/config/cuda.amazon.properties [hunk 1]
```diff
@@ -442,8 +442,16 @@ libs.cueigen.versions.337.version=3.3.7
 libs.cueigen.versions.337.path=/opt/compiler-explorer/libs/eigen/v3.3.7
 libs.cueigen.versions.334.version=3.3.4
 libs.cueigen.versions.334.path=/opt/compiler-explorer/libs/eigen/v3.3.4
-libs.thrustcub.name=Thrust+CUB
-libs.thrustcub.description=CUDA collective and parallel algorithms
+libs.cccl.name=CUDA C++ Core libraries
+libs.cccl.description=Fundamental CUDA C++ libraries: Thrust, CUB, libcudacxx
+libs.cccl.versions=trunk:2002000
+libs.cccl.url=http://www.github.com/NVIDIA/cccl
```
Comment (bdice) [inline etc/config/cuda.amazon.properties:448]: This is probably a more "normalized" URL:
```suggestion
libs.cccl.url=https://github.com/NVIDIA/cccl
```

File: etc/config/cuda.amazon.properties [hunk 2]
```diff
@@ -442,8 +442,16 @@ libs.cueigen.versions.337.version=3.3.7
...
+libs.cccl.versions.trunk.version=trunk
+libs.cccl.versions.trunk.path=/opt/compiler-explorer/libs/cccl/trunk/thrust:/opt/compiler-explorer/libs/cccl/trunk/cub:/opt/compiler-explorer/libs/cccl/trunk/libcudacxx/include
+libs.cccl.versions.2002000.version=2.2.0
+libs.cccl.versions.2002000.path=/opt/compiler-explorer/libs/cccl/2.2.0/thrust:/opt/compiler-explorer/libs/cccl/2.2.0/cub:/opt/compiler-explorer/libs/cccl/2.2.0/libcudacxx/include
+libs.thrustcub.name=Thrust+CUB (Legacy)
+libs.thrustcub.description= Use CUDA C++ Core Libraries (CCCL) for newer versions
```
Comment (bdice) [inline etc/config/cuda.amazon.properties:454]: Probably an extra space?
```suggestion
libs.thrustcub.description=Use CUDA C++ Core Libraries (CCCL) for newer versions
```

**Review by bdice — COMMENTED** [review]
Thanks for this, Jake!

File: etc/config/cuda.amazon.properties [hunk 3]
```diff
@@ -442,8 +442,16 @@ libs.cueigen.versions.337.version=3.3.7
...
 libs.cueigen.versions.334.path=/opt/compiler-explorer/libs/eigen/v3.3.4
-libs.thrustcub.name=Thrust+CUB
-libs.thrustcub.description=CUDA collective and parallel algorithms
+libs.cccl.name=CUDA C++ Core libraries
+libs.cccl.description=Fundamental CUDA C++ libraries: Thrust, CUB, libcudacxx
+libs.cccl.versions=trunk:2002000
+libs.cccl.url=https://github.com/NVIDIA/cccl
+libs.cccl.versions.trunk.version=trunk
+libs.cccl.versions.trunk.path=/opt/compiler-explorer/libs/cccl/trunk/thrust:/opt/compiler-explorer/libs/cccl/trunk/cub:/opt/compiler-explorer/libs/cccl/trunk/libcudacxx/include
+libs.cccl.versions.2002000.version=2.2.0
```
Comment (mattgodbolt) [inline etc/config/cuda.amazon.properties:451]: The ID is unimportant other than needing to be kept the same "forever", but I think `2002000` is confusing for 2.2.0. Or am I missing something?

Comment (partouf) [reply etc/config/cuda.amazon.properties:451]: I was wondering the same, I guess it's to pre-empt for possible 2.10.x where 10 is more than 2 and perhaps that it wouldn't sort correctly then? but we have semver ordering, so it should be fine if you just use id 220

Comment (mattgodbolt) [reply etc/config/cuda.amazon.properties:451]: other than that the paths look good!

**Review by mattgodbolt — COMMENTED** [review]
Will let you get back to us on the ID but if not I'll merge on Monday anyway

File: etc/config/cuda.amazon.properties (hunk 3 above)
Comment (jrhemstad) [reply etc/config/cuda.amazon.properties]: I updated it to `220`.

**Comment by mattgodbolt** [comment]
Thanks! Will merge shortly :)

**Comment by jrhemstad** [comment]
Hey @mattgodbolt @partouf, I must have done something wrong, because using the new library isn't adding the appropriate include paths to the compile line. 

To test, I tried printing the version of Thrust using the new CCCL library and the old Thrust+CUB library. Using the new CCCL library shows the old version (meaning it's using the version bundled with nvcc instead). Whereas the old Thrust+CUB library is showing the correct 200200 version. 

More importantly, looking at the compiler options:

CCCL
> -o /tmp/compiler-explorer-compiler2023819-11214-ksezdo.4sra/output.s -g -lineinfo --keep-device-functions -c -keep -keep-dir /tmp/compiler-explorer-compiler2023819-11214-ksezdo.4sra -Xcompiler=-S --compiler-bindir /opt/compiler-explorer/gcc-10.2.0/bin /tmp/compiler-explorer-compiler2023819-11214-ksezdo.4sra/example.cu


Thrust+CUB

> -o /tmp/compiler-explorer-compiler2023819-10590-1vtb5o3.fynp/output.s -g -lineinfo --keep-device-functions -c -keep -keep-dir /tmp/compiler-explorer-compiler2023819-10590-1vtb5o3.fynp -Xcompiler=-S --compiler-bindir /opt/compiler-explorer/gcc-10.2.0/bin /tmp/compiler-explorer-compiler2023819-10590-1vtb5o3.fynp/example.cu **-I/opt/compiler-explorer/libs/thrustcub/trunk -I/opt/compiler-explorer/libs/thrustcub/trunk/dependencies/cub**

The equivalent include path options are missing. Did I do something wrong when specifying the include paths here? https://github.com/compiler-explorer/compiler-explorer/blob/12260ed81076ec2e449d5eeeedec70871ae22ded/etc/config/cuda.amazon.properties#L450


https://godbolt.org/z/Yfd6esqbo

**Comment by partouf** [comment]
this isn't actually live yet

**Comment by partouf** [comment]
I think this is fixed now

**Pull request merged**